Cognito and JWT

GiessC opened the pull request.

GiessC committed:
Web Posts API, working on form, broken rn

GiessC committed:
Axios issues importing base url from .env

GiessC committed:
Working on API hooks for web app, errors tho

GiessC committed:
Posts API for web

GiessC committed:
Finished posts integration with web

GiessC committed:
Cleanup

GiessC committed:
Nice

GiessC committed:
AuthService setup for Cognito

GiessC committed:
User sign up works, working on confirmation, login may work

GiessC committed:
Having trouble with PostConfirmationTrigger

GiessC committed:
Fixed PostConfirmationTrigger so it properly adds to DynamoDB, login, confirmation, and logout works

GiessC commented on infrastructure/bin/infrastructure.ts:
```diff
@@ -1,12 +1,12 @@
-new CognitoStack(app, 'sweep-cognito-stack');
-new DynamoStack(app, 'sweep-dynamo-stack');
+const dynamoStack = new DynamoStack(app, 'sweep-dynamo-stack');
+new CognitoStack(app, 'sweep-cognito-stack', dynamoStack.usersTable);
```
I don't love directly passing the Users table in here. It makes it easy to add the correct permissions, but it might be better to handle that ourselves anyways. I think we should have the CDK output the ARN of the Users table for the CognitoStack to use.

GiessC committed:
Broken infra changes

GiessC committed:
Progress

GiessC committed:
User can now forget and reset password

GiessC commented:
Unrelated to my most recent changes, but the infrastructure is creating two app clients for the Cognito user pool, even though we only specify one. The CDK must be automatically creating one for a reason unbeknownst to me.

GiessC committed:
Added error handling for all auth

GiessC committed:
Working with JWT on API

GiessC committed:
JWT implemented

GiessC committed:
Post author inherited from JWT token

GiessC committed:
Refresh token on fetch

GiessC committed:
Refreshes token before logout

GiessC committed:
Renamed stacks and removed second cognito app client

LeTomJones commented on api/src/app.ts:
```diff
@@ -35,16 +14,7 @@ if (process.env.NODE_ENV == 'local') {
-};
-
-export const database = Database.getInstance(dbConfig);
+export const postgres = setupPostgres();
```
This approach, where you have extracted the major functionality into setup functions, is much better

LeTomJones commented on api/src/features/posts/endpoints/CreatePostEndpoint.ts:
```diff
@@ -0,0 +1,64 @@
+        };
+        response.status(StatusCodes.UNAUTHORIZED).send(body);
+        return;
+    }
```
Instead of duplicating your efforts here, I think it makes more sense to just return if these are falsy. The middleware will handle responding with errors already.

This happens on many of the endpoints, so I'm going to use `Ref-1` to refer to this same problem.

LeTomJones commented on api/src/features/posts/endpoints/CreatePostEndpoint.ts:
```diff
@@ -0,0 +1,64 @@
+        };
+        response.status(StatusCodes.UNAUTHORIZED).send(body);
+        return;
+    }
```
`Ref-1`

LeTomJones commented on api/src/features/posts/endpoints/CreatePostEndpoint.ts:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    const repository: IPostRepository = PostRepository.getInstance(
+        getPostDBProvider(),
```
I haven't checked yet, but if this is called every time we get the PostRepository instance - remove it from the parameters and just call it in the getInstance() method.

LeTomJones commented on api/src/features/posts/endpoints/DeletePostEndpoint.ts:
```diff
@@ -0,0 +1,33 @@
+        };
+        response.status(StatusCodes.BAD_REQUEST).send(body);
+        return;
+    }
```
`Ref-1`

LeTomJones commented on api/src/features/posts/endpoints/EditPostEndpoint.ts:
```diff
@@ -0,0 +1,46 @@
+        };
+        response.status(StatusCodes.BAD_REQUEST).send(body);
+        return;
+    }
```
`Ref-1`

LeTomJones commented on api/src/features/posts/endpoints/EditPostEndpoint.ts:
```diff
@@ -0,0 +1,46 @@
+        };
+        response.status(StatusCodes.BAD_REQUEST).send(body);
+        return;
+    }
```
The error handling here isn't great. I don't like assuming it's a bad request just because there is no post. Use a try-catch instead.

LeTomJones commented on api/src/features/posts/endpoints/CreatePostEndpoint.ts:
```diff
@@ -0,0 +1,64 @@
+        content: requestBody.content,
+        author: username,
+        authorId: userId,
+    });
```
Add error handling using try-catch. You can use the [DynamoDB Documentation](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_PutItem.html#API_PutItem_Errors) to find the possible errors that will occur. 

1. Add a try-catch to the PostDynamoDBProvider methods to catch DynamoDB errors and return one of our own (which you will have to create). 
2. Add a try-catch to the calls from endpoints to the repository to catch "our" errors
3. If there are any errors, return them in the response

Do this on all endpoints, but I'm not writing this comment again lol

LeTomJones commented on api/src/features/posts/endpoints/GetPostEndpoint.ts:
```diff
@@ -0,0 +1,50 @@
+        };
+        response.status(StatusCodes.BAD_REQUEST).send(body);
+        return;
+    }
```
I'm noticing we do this on every endpoint handler. Let's extract it somewhere by making a function that takes in the request and response objects and will send a response if there are validation issues with the request.

LeTomJones commented on api/src/features/posts/middleware/canDeletePost.ts:
```diff
@@ -0,0 +1,57 @@
+        };
+        response.status(StatusCodes.UNAUTHORIZED).send(body);
+        return;
+    }
```
`Ref-1`

LeTomJones commented on api/src/features/posts/middleware/canDeletePost.ts:
```diff
@@ -0,0 +1,57 @@
+        };
+        response.status(StatusCodes.UNAUTHORIZED).send(body);
+        return;
+    }
```
`Ref-1`

LeTomJones commented on api/src/features/posts/middleware/canEditPost.ts:
```diff
@@ -0,0 +1,51 @@
+        };
+        response.status(StatusCodes.UNAUTHORIZED).send(body);
+        return;
+    }
```
`Ref-1`

LeTomJones commented on api/src/features/posts/middleware/canEditPost.ts:
```diff
@@ -0,0 +1,51 @@
+        };
+        response.status(StatusCodes.UNAUTHORIZED).send(body);
+        return;
+    }
```
`Ref-1`

LeTomJones commented on api/src/features/users/models/dto/UserDto.ts:
```diff
@@ -8,12 +8,19 @@ export default class UserDto {
     public readonly sk: string;
     public readonly id: string;
     public readonly username: string;
+    public readonly roles: string[];
```
Make a Role enum instead of using strings

LeTomJones commented on api/src/middleware/jwt.ts:
```diff
@@ -0,0 +1,43 @@
+) => {
+    const token = getTokenFromHeaders(request.headers);
+    if (!token) {
+        response.status(StatusCodes.UNAUTHORIZED).send('Unauthorized');
```
Make this message a little more descriptive, something like 'You are not authorized to perform this action.'

LeTomJones commented on api/src/setup/postgres/PostgresSetup.ts:
```diff
@@ -0,0 +1,16 @@
+    return Database.getInstance(dbConfig);
+};
+
+export default setupPostgres;
```
This stuff won't be used, but removing it is outside of the scope of this PR anyways. Will be removed in issue #27

LeTomJones commented on infrastructure/lib/CognitoStack.ts:
```diff
@@ -0,0 +1,164 @@
+import { Runtime } from 'aws-cdk-lib/aws-lambda';
+import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
+import { Construct } from 'constructs';
+import path = require('path');
```
Are you sure you have to use require here?

LeTomJones commented on infrastructure/lib/lambda/PostConfirmationTriggerCognito.ts:
```diff
@@ -26,10 +26,14 @@ export const handler = async (event: PostConfirmationTriggerEvent) => {
     } catch (error: unknown) {
         console.error(error);
+    } finally {
+        return event;
```
Don't do this in a finally block, just return it and remove the finally. There are some weird side effects that occur, so you should never perform jumps in finally blocks.

LeTomJones commented on infrastructure/lib/lambda/PostConfirmationTriggerCognito.ts:
```diff
@@ -26,10 +26,14 @@ export const handler = async (event: PostConfirmationTriggerEvent) => {
             }),
         );
     } catch (error: unknown) {
         console.error(error);
```
I think you want to rethrow the error

LeTomJones commented on infrastructure/lib/lambda/PreTokenGenerationTriggerCognito.ts:
```diff
@@ -0,0 +1,69 @@
+    } catch (error: unknown) {
+        if (error instanceof ResourceNotFoundException) {
+            return undefined;
+        } else {
```
Remove this else, it's unnecessary

LeTomJones commented on web/.env.template:
```diff
@@ -1,6 +0,0 @@
-AUTH0_CLIENT_SECRET="" # Get these from auth0
-AUTH0_BASE_URL="" # This is our domain
-AUTH0_SECRET="" # Generate using `openssl rand -hex 32`
-AUTH0_CALLBACK="" # This is a relative URL
```
There should probably be another issue for removing the Auth0 stuff.

LeTomJones commented on web/src/api/utils/fetch.ts:
```diff
@@ -1,70 +1,83 @@
+    };
+    const user = AuthService.getInstance().user;
+    if (user && tokenExpired(user)) {
+        await AuthService.getInstance().refreshIdToken();
```
I don't know if the `user` variable will be updated

LeTomJones commented on web/src/components/Navbar.tsx:
```diff
@@ -1,42 +1,127 @@
+    const { isAuthenticated } = useContext<IAuthContext>(AuthContext);
     const [logoutModalOpen, setLogoutModalOpen] = useState<boolean>(false);
-    const { user } = useUser();
+    const [anchorEl, setAnchorEl] = useState<Element | null>(null);
```
Let's use a more specific name for this.

LeTomJones commented on web/src/components/Navbar.tsx:
```diff
@@ -1,42 +1,127 @@
+        // TODO: Create a navigation menu for a user's communities and such.
+    };
+
+    const handleUserMenu = (event: React.MouseEvent<HTMLButtonElement>) => {
```
Use on, not handle.

LeTomJones commented on web/src/components/Navbar.tsx:
```diff
@@ -1,42 +1,127 @@
 
 const Navbar = () => {
+    const router = useRouter();
+    const { isAuthenticated } = useContext<IAuthContext>(AuthContext);
```
isAuthenticated doesn't exist

LeTomJones commented on web/src/components/auth/confirm-user/ConfirmUserForm.tsx:
```diff
@@ -0,0 +1,142 @@
+            router.push('/auth/login');
+            return;
+        } catch (error: unknown) {
+            if (isAWSError(error, 'CodeMismatchException')) {
```
I think we can just check if it's an instance of CodeMismatchException, but I'm not sure if these are provided for TypeScript.

LeTomJones commented on web/src/components/auth/confirm-user/ConfirmUserForm.tsx:
```diff
@@ -0,0 +1,142 @@
+                    enter it here.
+                </Typography>
+                <FormGroup className='mt-4'>
+                    <TextField
```
Eventually, we will want to create a better component for this, which a textbox for each number, but this works for now.

GiessC replied:
Created an issue for this #54

LeTomJones commented on web/src/components/auth/password/forgot/ForgotPasswordForm.tsx:
```diff
@@ -0,0 +1,154 @@
+        <Box className='h-full'>
+            <Typography variant='h4'>Forgot Password</Typography>
+            <Typography variant='body1'>
+                Please provide your email and username. We will send you an
```
Change this to username and email, since that's the order of the TextFields in the form.

LeTomJones commented on web/src/errors/ErrorMessages.ts:
```diff
@@ -0,0 +1,31 @@
+export const NOT_FOUND = (friendlyName: string, searchKey: string) =>
+    `No ${friendlyName.toLowerCase()} found with the given ${searchKey}.`;
+
+export const INVALID_EMAIL = () => 'Please specify a valid email';
```
Add a period at the end.

LeTomJones commented on web/src/features/auth/confirm-user/schema.ts:
```diff
@@ -0,0 +1,8 @@
+import * as Yup from 'yup';
+
+const confirmUserSchema = Yup.object({
+    code: Yup.string().required('Code is required'),
```
Use the REQUIRED function from ErrorMessages

GiessC committed:
Changed unauthorized message and post will get token using context not AuthService

GiessC commented:
> Unrelated to my most recent changes, but the infrastructure is creating two app clients for the Cognito user pool, even though we only specify one. The CDK must be automatically creating one for a reason unbeknownst to me.

Fixed

GiessC committed:
Added try-catch to findAll for PostDynamoDBProvider, some other small changes

GiessC committed:
Mostly extraction, error handling, and small fixes

GiessC commented:
I've implemented all changes that were requested, but as I've been testing the app, I've run into problems with detecting whether a user is authorized. On each page refresh, the user is reset to null. It may have to be stored in the AuthContext, but I don't have time to test it right now.

Since the user is set to null on each refresh, the API returns an `Unauthorized` response on every request and the Navbar component doesn't update the `Login` text to the user menu.

GiessC commented:
> I've implemented all changes that were requested, but as I've been testing the app, I've run into problems with detecting whether a user is authorized. On each page refresh, the user is reset to null. It may have to be stored in the AuthContext, but I don't have time to test it right now.
> 
> Since the user is set to null on each refresh, the API returns an `Unauthorized` response on every request and the Navbar component doesn't update the `Login` text to the user menu.

Fixed! I lied and stayed up nice and late to fix the errors. Got rid of the useAuth() hook and instead put our context value inside of our AuthProvider. Everything works again!

GiessC committed:
All PR change requests, removed useAuth and moved to AuthProvider

GiessC committed:
Fixed error handling for endpoints, removed dumb ErrorHandler service
